### PR TITLE
Specialise hs

### PR DESCRIPTION
**Pull request by andrevidela**
This PR contains specialisation for haskell types. 

Specialisation is implemented using a `SortedMap` which maps Strings to `HsType`. Those `HsType` are the haskell representation of existing types.

Those types are not serialised or deserialised _yet_ but they will generate correct type signatures using existing specialized types.

Though the function `extendContext` allows to transform a `TDef n` (with free variables) to a `TDefR n` it isn't used for the haskell backend since the current implementation works. I will issue another PR cleaning up the Haskell backend using the total version of `makeType` rather than the partial one.

Update: as we've seen with @clayrat implementing `extendContext` is non-trivial regarding type application. we will see what transpires from that later

### CODE REVIEW

**Comment by andrevidela** [comment]
alright gotta clean up the history but you can already review the diff @clayrat @fredrikNordvallForsberg

File: src/Typedefs/Test/SpecialisationTests.idr [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+eitherBoolInt = TName "A" (TSum [FRef "Bool" , FRef "Int" ])
+
+testSuite : IO ()
+testSuite = spec $ do
+  describe "specialisation tests" $ do
+    it "should use Bools" $
+      ((unwords . map (toString . renderDef)) <$> generateTyDefs {def=Haskell} {type=HsType} (MkNEList (Unbounded eitherBool) []))
+        `shouldBe`
+      (pure "type A = Either Bool Bool")
+    it "should aosudhf" $
```
Comment (clayrat) [inline src/Typedefs/Test/SpecialisationTests.idr]: Something meaningful here would also be nice :)

Comment (andrevidela) [reply src/Typedefs/Test/SpecialisationTests.idr]: fixed

**Comment by andrevidela** [comment]
I reviewed your changes and they seem sensible. I ran the tests locally and they compile, not sure why CI is screaming as usual.

File: src/Typedefs/Backend.idr [hunk 2]
```diff
@@ -12,18 +12,18 @@ import Text.PrettyPrint.WL
...
   Zero : p Z -> ZeroOrUnbounded p False
 
-fromSigma : {p : Nat -> Type} -> (b : Bool) -> (n ** p n) -> Maybe (ZeroOrUnbounded p b)
-fromSigma True  (n  **pn) = Just $ Unbounded $ pn
-fromSigma False (Z  **pz) = Just $ Zero $ pz
-fromSigma False (S _** _) = Nothing
+fromSigma : {p : Nat -> Type} -> (b : Bool) -> (n ** p n) -> Either CompilerError (ZeroOrUnbounded p b)
+fromSigma True  (n  **pn) = Right $ Unbounded $ pn
+fromSigma False (Z  **pz) = Right $ Zero $ pz
+fromSigma False (S _** _) = Left $ UnknownError "Inconsisten sigma"
```
[the remark this answers is not in the document]
Comment (andrevidela) [reply src/Typedefs/Backend.idr]: ah good point I was also confused by this at first

File: src/Typedefs/Backend/Specialisation.idr [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+interface Specialisation t where
+  specialisedTypes : SortedMap String t
+
+sumVect : Vect n (k ** Vect k a) -> (j ** Vect j a)
+sumVect [] = (Z ** [])
+sumVect ((k ** ks) :: xs) with (sumVect xs)
+  | (j ** js) = (k + j ** ks ++ js)
+
+||| extract all the references from a TDef
+allRefs : (TDef n) -> (m ** Vect m Name)
```
[the remark this answers is not in the document]
Comment (andrevidela) [reply src/Typedefs/Backend/Specialisation.idr]: good catch, I removed it and used `getUsedIndices` instead

File: src/Typedefs/Backend/Specialisation.idr [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+weakenSN : (m : Nat) -> Fin (S n) -> Fin (S (n + m))
+weakenSN m FZ = FZ
+weakenSN m (FS x) = weaken $ weakenN m x
+
+pairLookup : k -> SortedMap k v -> Either k (k, v)
+pairLookup k m = MkPair k <$> maybeToEither k (lookup k m)
+
+||| Build a vector of all the references, without duplication, present in the SortedMap
+buildVect : (t : TDef n) -> SortedMap String b -> Either String (l ** Vect l (String, b))
```
[the remark this answers is not in the document]
Comment (andrevidela) [reply src/Typedefs/Backend/Specialisation.idr]: renamed to `collectReferences`

**Comment by andrevidela** [comment]
Should be done! I've patched up the tests for specialisation but I'm unsure about CI being able to run the tests. Anyone knows what's up?

**Comment by andrevidela** [comment]
Disabling specialisation tests fixes the timeout. Since they apparently work for both @epost and me locally I'm going to assume they aren't broken but exhibit some pathological behaviour on travis. I've opened an issue to take care of this later. #202

**Pull request merged**